contrib/gideal/gideal.man
.\" t
.TH GIDEAL @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
gideal \(em \fIgroff\fP preprocessor for the \fIideal\fP language
.
.\" The .SH was moved to this place in order to appease `apropos'.
.
.\" gideal.1 - man page for the ideal roff preprocessor
.\" Source file position:  <groff_source_top>/contrib/gideal/gideal.man
.\" Installed position:    $prefix/share/man/man1/gideal.1
.
.
.\" --------------------------------------------------------------------
.\" Legalese
.\" --------------------------------------------------------------------
.
.de co
Copyright \[co] 2014 Free Software Foundation, Inc.

This file is part of gideal, which is part of groff, a free software
project.

You can redistribute it and/or modify it under the terms of the GNU
General Public License version 2 (GPL2), as published by the Free
Software Foundation.

The license text is available in the internet at
.UR http://\:www.gnu.org/\:licenses/\:gpl-2.0.html
.UE .
..
.
.de au
This file was written by
.MT bernd.warken@web.de
Bernd Warken
.ME .
..
.
.\" --------------------------------------------------------------------
.\" Characters
.\" --------------------------------------------------------------------
.
.\" Ellipsis ...
.ie t .ds EL \fS\N'188'\fP
.el .ds EL \&.\|.\|.\&\
.\" called with \*(EL
.
.\" Bullet
.ie t .ds BU \(bu
.el .ds BU *
.\" used in `.IP \*(BU 2m' (former .Topic)
.
.
.\" --------------------------------------------------------------------
.SH SYNOPSIS
.\" --------------------------------------------------------------------
.
.SY gideal
.OP \-
.OP \-\-
.OP \fI\%filespec \*(EL
.YS
.
.BR "gideal \-h" | "\-\-help"
.br
.BR "gideal \-v" | "\-\-version"
.
.
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.\" --------------------------------------------------------------------
.
.P
.B gideal
is a
.I groff preprocessor
for using the roff language
.BR ideal (@MAN7EXT@)
for drawing pictures within
.I groff files
(to be done).
.
.
.\" --------------------------------------------------------------------
.SH OPTIONS
.\" --------------------------------------------------------------------
.
In
.BR gideal ,
there are the following kinds of options:
.IP \*(BU 2m
.I breaking options
.IP \*(BU 2m
.I filespec options
.
.
.\" --------------------------------------------------------------------
.SS "Breaking Options"
.\" --------------------------------------------------------------------
.
.I Breaking options
are used to display a corresponding information, but then the whole
program run is stopped and finished.
.
.
.P
The breaking options
.B \-v
or
.B \-\-version
print information on the version number.
.
.
.P
The breaking options
.B \-h
or
.B \-\-help
print usage information.
.
.
.P
Both of these options automatically end the program.
.
Other options are then ignored.
.
.
.\" --------------------------------------------------------------------
.SS "Filespec Options"
.\" --------------------------------------------------------------------
.
Each argument that does not start with a minus
.B \-
character is interpreted as a file name.
.
.
.P
As usual, the single minus character argument
.B \-
means to integrate
.IR "standard input" .
.
.
.P
The double minus
.B \-\-
argument means noting special by itself, but means that all following
arguments will be interpreted as file names, even if it starts with a
minus
.B \-
character.
.
.
.\" --------------------------------------------------------------------
.SH HISTORY
.\" --------------------------------------------------------------------
.
The roff language
.BR ideal (@MAN7EXT@)
was invented by
.I Christopher J. Wyk
in 1981 at AT&T.
.
.
.P
Moreover, a roff preprocessor for this language with the name
.B ideal
was invented by the same programmer as well, but this preprocessor
looks now quite old-fashioned.

.
There is now a complitely new preprocessor for the groff system in
development.
.
It will be called
.BR gideal (@MAN1EXT@),
other than the original preprocessor.
.
.
.\" --------------------------------------------------------------------
.SH "SEE ALSO"
.\" --------------------------------------------------------------------
.
.TP
.B Unix8 preprocessor
freely available documentation for the
.IR "ideal preprocessor" .
.
The
.I preprocessor
there is called
.BR ideal ,
not
.B gideal
as in
.IR groff .
.br
.UR http://man.cat-v.org/unix_8th/1/ideal
.UE .
.
.
.\" --------------------------------------------------------------------
.SH "COPYING"
.\" --------------------------------------------------------------------
.co
.\" --------------------------------------------------------------------
.SH "AUTHORS"
.\" --------------------------------------------------------------------
.au
.
.
.\" --------------------------------------------------------------------
.\" Emacs settings
.\" --------------------------------------------------------------------
.
.\" Local Variables:
.\" mode: nroff
.\" End:


contrib/gideal/gideal.pl
#!/usr/bin/env perl

# groff ideal

# Source file position: <groff-source>/contrib/gideal/gideal.pl
# Installed position: <prefix>/bin/gideal

# Copyright (C) 2014  Free Software Foundation, Inc.

# Written by Bernd Warken <groff-bernd.warken-72@web.de>.

my $version = '0.9.9';

# This file is part of `gideal', which is part of `groff'.

# `groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

# `groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see
# <http://www.gnu.org/licenses/gpl-2.0.html>.

########################################################################

use strict;
use warnings;

# temporary dir and files
use File::Temp qw/ tempfile tempdir /;

# needed for temporary dir
use File::Spec;

# for `copy' and `move'
use File::Copy;

# for fileparse, dirname and basename
use File::Basename;

# current working directory
use Cwd;

# $Bin is the directory where this script is located
use FindBin;

# for running the ideal parts with the `IS' and `I[EF]' programs and
# getting output also useful for shell programs
use IPC::System::Simple qw(capture capturex run runx system systemx);

# Perl package for complex numbers
use Math::Complex;


########################################################################
# system variables
########################################################################

$\ = "\n";    # adds newline at each print
$/ = "\n";    # newline separates input
$| = 1;       # flush after each print or write command


########################################################################
# read-only variables with double-@ construct
########################################################################

my $before_make;	# script before run of `make'
{
  my $at = '@';
  $before_make = 1 if '@VERSION@' eq "${at}VERSION${at}";
}

my %at_at;
my $gideal_dir;

if ($before_make) {
  my $gideal_source_dir = $FindBin::Bin;
  $at_at{'BINDIR'} = $gideal_source_dir;
  $at_at{'G'} = '';
} else {
  $at_at{'BINDIR'} = '@BINDIR@';
  $at_at{'G'} = '@g@';
}


########################################################################
# options
########################################################################

foreach (@ARGV) {
  if ( /^(-h|--h|--he|--hel|--help)$/ ) {
    print q(Usage for the `gideal' program:);
    print 'gideal [-] [--] [filespec...] normal file name arguments';
    print 'gideal [-h|--help]            gives usage information';
    print 'gideal [-v|--version]         displays the version number';
    print q(This program is a `groff' preprocessor that handles ideal ) .
      q(parts in `roff' files.);
    exit;
  } elsif ( /^(-v|--v|--ve|--ver|--vers|--versi|--versio|--version)$/ ) {
    print q(`gideal' version ) . $version;
    exit;
  }
}


#######################################################################
# temporary file
#######################################################################

my $out_file;
{
  my $template = 'gideal_' . "$$" . '_XXXX';
  my $tmpdir;
  foreach ($ENV{'GROFF_TMPDIR'}, $ENV{'TMPDIR'}, $ENV{'TMP'}, $ENV{'TEMP'},
	   $ENV{'TEMPDIR'}, 'tmp', $ENV{'HOME'},
	   File::Spec->catfile($ENV{'HOME'}, 'tmp')) {
    if ($_ && -d $_ && -w $_) {
      eval { $tmpdir = tempdir( $template,
				CLEANUP => 1, DIR => "$_" ); };
      last if $tmpdir;
    }
  }
  $out_file = File::Spec->catfile($tmpdir, $template);
}


########################################################################
# input
########################################################################

my $nr = 0;
my @ideal;

my $ideal_mode = 0;
foreach (<>) {
  chomp;
  my $line = $_;
  my $is_start = $line =~ /^[.']\s*IS(|\s+.*)$/;	# start .IS
  my $is_end = $line =~ /^[.']\s*(I[EF])(|\s+.*)$/;	# stop .IE/.IF

  # .IE below ideal picture (after), .IF back at starting point (before)
  my $end_mode = $1;

  if ( $is_start ) {
    if ( $ideal_mode ) {
      print STDERR
	'Within a running ideal part, you called another .IS!';
      next;
    } else {	# new start
      $ideal_mode = 1;
      open OUT, '>', $out_file;
      @ideal = ();
      next;
    }
  }

  if ( $is_end ) {
    $ideal_mode = 0;
    close OUT;
    &do_ideal();
    next;
  }

  if ( $ideal_mode ) {	# normal `ideal' line, store in @ideal
    push @ideal, $line;
    next;
  } else {		# not related to `ideal', so print to STDOUT
    print $line;
    next;
  }
  next;
}


########################################################################
# &do_ideal()
########################################################################

sub do_ideal {
    $nr++;
    print '### ideal part number ' . $nr . ':';
    foreach (@ideal) {
      print $nr . '# ' . $_;
    }
    @ideal = ();
} # &do_ideal()


1;
########################################################################
### Emacs settings
# Local Variables:
# mode: CPerl
# End:
